File: session-2/400.cpp
// 400. Nth Digit
#include "leetcode.hpp"
using namespace std;

class Solution {
public:
    int findNthDigit(int n) {
        int level = 1;
        int level_cnt = (pow(10, level) - pow(10, level - 1)) * level;

        while (n - level_cnt > 0) {
            n -= level_cnt;
            level += 1;
            level_cnt = (pow(10, level) - pow(10, level - 1)) * level;
        }

        int order = (n % level == 0) ? (n / level) : (n / level + 1);
        int number = pow(10, level - 1) + order - 1;
        int digit_position = (n % level == 0) ? level : (n % level);

        digit_position = level - digit_position;
        return number / (int)pow(10, digit_position) % 10;
    }
};

int main() {
    Solution s;
    for (int i = 1; i <= 1000; ++i)
        cout << s.findNthDigit(i);
    // cout << s.findNthDigit(100) << endl;
}

File: leetcode/session-2/400.java
/*
 * 400. Nth Digit
 */
public class Solution {
    public int findNthDigit(int n) {
        long numBits = 1;
        long numCnts = 9;
        long base = 1;

        long product = numBits * numCnts;
        while (n > product) {
            n -= product;
            numBits += 1;
            numCnts *= 10;
            base *= 10;
            product = numBits * numCnts;
        }

        long offset = n / numBits;
        long digitOffset = n % numBits;
        long number = base + (digitOffset > 0 ? offset + 1 : offset) - 1;
        if (numBits == 1) return (int)number;

        long res = 0;
        long cirCnt = digitOffset == 0 ? 0 : numBits - digitOffset;
        for (int i = 0; i <= cirCnt; ++i) {
            res = number % 10;
            number /= 10;
        }
        return (int)res;
    }
}
